File: Scheduler/src/alma/SCHEDULER_MODULE/SchedulerImpl/SchedulerImpl.java
package alma.SCHEDULER_MODULE.SchedulerImpl;

import alma.acs.component.ComponentLifecycle;
import alma.acs.component.ComponentLifecycleException;
import alma.acs.container.ContainerServices;
import java.util.logging.Logger;
import alma.ACS.ComponentStates;
import alma.JavaContainerError.wrappers.AcsJContainerServicesEx;
import alma.SCHEDULER_MODULE.SchedulerOperations;
import alma.DATABASE_MODULE.DataBase;
import alma.INSTRUMENT_MODULE.Instrument;
import alma.TELESCOPE_MODULE.Telescope;
import alma.TYPES.Proposal;
import alma.TYPES.Target;




public class SchedulerImpl implements ComponentLifecycle, SchedulerOperations {
	
	
	private ContainerServices m_containerServices;
    private Logger m_logger;
    static private int PROPOSAL_QUEUED=0;
    static private int PROPOSAL_RUNNING=1;
    static private int PROPOSAL_READY=2;
    private TheLoop loop;
    boolean loopValidator = true;
    boolean startValidator = false;
    int currentpid = -1;
    
    
    private class TheLoop extends Thread {
    	
    	
    	public void run() {
    		startValidator = true;
    		
    		Proposal prToRun;
    		org.omg.CORBA.Object compObj;
    		try {
    			compObj = m_containerServices.getComponent("DATABASE");
    			//assertNotNull(compObj);
    			DataBase dbComp = alma.DATABASE_MODULE.DataBaseHelper.narrow(compObj);
    			
    			compObj = m_containerServices.getComponent("INSTRUMENT");
    			//assertNotNull(compObj);
    			Instrument instrComp = alma.INSTRUMENT_MODULE.InstrumentHelper.narrow(compObj);
    			
    			compObj = m_containerServices.getComponent("TELESCOPE");
    			//assertNotNull(compObj);
    			Telescope telComp = alma.TELESCOPE_MODULE.TelescopeHelper.narrow(compObj);
    			
    			while(loopValidator){
    				m_logger.info("getProposals");
    				Proposal[] proposals = dbComp.getProposals();
    				
    				prToRun = null;
    				for (Proposal proposal : proposals) {	
    					if(proposal.status == PROPOSAL_QUEUED){
    						prToRun = proposal;
    						break;
    					}
    				}
    				
    				if(prToRun != null){
    					
    					currentpid = prToRun.pid;
    				
    					m_logger.info("setProposalStatus running");
    					dbComp.setProposalStatus(prToRun.pid,PROPOSAL_RUNNING);
    					m_logger.info("CameraOn");
    					instrComp.cameraOn();
    					Target[] targets = prToRun.targets;
    					for (Target target : targets) {
    						m_logger.info("storeImage");
    						dbComp.storeImage(prToRun.pid, target.tid, telComp.observe(target.coordinates, target.expTime));
    						}
    					m_logger.info("CameraOff");
    					instrComp.cameraOff();
    					m_logger.info("setProposalStatus ready");
    					dbComp.setProposalStatus(prToRun.pid,PROPOSAL_READY);
    				} else {
    					try {
    						m_logger.info("sleeping for 2 seconds...");
							Thread.sleep(2000);
							currentpid = -1;
						} catch (InterruptedException e) {
							// TODO Auto-generated catch block
							e.printStackTrace();
						}
    				}
    			}
    			if(loopValidator == false){
    				startValidator = false;
    				currentpid = -1;
    			}
    			
    			
    		} catch (AcsJContainerServicesEx e) {
    			// TODO Auto-generated catch block
    			e.printStackTrace();
    		}
    		
    	}
  
    }
    
    
    
    
/**
 * Implementation of ComponentLifecycle
 */
    
	public void initialize(ContainerServices containerServices) throws ComponentLifecycleException {
		// TODO Auto-generated method stub
		m_containerServices = containerServices;
        m_logger = m_containerServices.getLogger();
        m_logger.info("initialize() called...");

	}
    
	public void execute() throws ComponentLifecycleException {
		// TODO Auto-generated method stub
		m_logger.info("execute() called...");

	}
	
	public void cleanUp() {
		// TODO Auto-generated method stub
		m_logger.info("cleanUp() called..., nothing to clean up.");

	}
	
	public void aboutToAbort() {
		// TODO Auto-generated method stub
		
	}
	
   /**
    * Implementation of ACSComponent
    */

	
	public ComponentStates componentState() {
		// TODO Auto-generated method stub
		return m_containerServices.getComponentStateManager().getCurrentState();
	}

	public String name() {
		return m_containerServices.getName();
	}

	
	
    /**
     * Implementation of SchedulerOperations
     */

	
	/*
     * Start the scheduler.
     * The scheduler will loop through all available proposals,
     * either until all proposals are done or until the stop method is called.
     * Raises an exception if called twice.
     * @return None
     */
    public void start (){
    	m_logger.info("Initialized ScheduleImpl.start()");
    	
    	
    	if(startValidator == false){
    		loopValidator = true;
        	loop = new TheLoop();
        	loop.start();
    	}
    	else{
    		m_logger.warning("Start() method already in use! :P");
    	}

    }
    //          raises(UOSErr::SchedulerAlreadyRunningEx);

    /*
     * Stop the scheduler.
     * This will stop the scheduler from scheduling more proposals,
     * it will not(!) break the ongoing observation.
     *
     * Raises an exception if called twice.
     * @return None
     */
    public void stop (){
    	m_logger.info("Initialized ScheduleImpl.stop()");
    	loopValidator = false;
    	if(startValidator){
    		try {
    			loop.join();
    		} catch (InterruptedException e) {
    			// TODO Auto-generated catch block
    			e.printStackTrace();
    		}
    	}else{
		m_logger.warning("Start() method must be used first.");
    	}
    	
    }
    //          raises(UOSErr::SchedulerAlreadyStoppedEx);

    /**
     * Returns the pid of the proposal currently under execution
     *
     * Raises exception if no proposal is executing.
     * @retun Proposal ID
     */
    public int proposalUnderExecution(){
    	m_logger.info("current pid = " + currentpid);
		return currentpid;
    }
//          raises(UOSErr::NoProposalExecutingEx);









	
}

File: SCHEDULERCPP1/src/SchedulerImpl.cpp
#include "SchedulerImpl.h"
#include "SchedulerThread.h"
#include <TypesS.h>
#include <SYSTEMErr.h>

const int SchedulerImpl::QUEUED = 0;
const int SchedulerImpl::RUNNING = 1;
const int SchedulerImpl::READY = 2;

SchedulerImpl::SchedulerImpl(const ACE_CString &name, maci::ContainerServices *cs)
: ACSComponentImpl(name, cs),
m_controlLoop_p(0),
pid_m(-1)
{
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::SchedulerImpl"));
}

SchedulerImpl::~SchedulerImpl()
{
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::~SchedulerImpl"));
	// destroy the thread via ThreadManager
	if(m_controlLoop_p!=0)
		getContainerServices()->getThreadManager()->destroy(m_controlLoop_p);
}

void SchedulerImpl::start() throw (SYSTEMErr::SchedulerAlreadyRunningEx)
{
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::start"));

	// starts the thread
	// if the scheduler has never been started before, we start the thread.  otherwise the 
 	// scheduler raise an exception.
 	if(m_controlLoop_p == 0)
	{
		ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::start creating a thread"));
		SchedulerImpl* selfPtr = this;
		m_controlLoop_p = getContainerServices()->getThreadManager()->
			create<SchedulerThread, SchedulerImpl*>("schedulerControlLoop", selfPtr);
		m_controlLoop_p->setSleepTime(10000000);
                m_controlLoop_p->resume();
	}
	else if(m_controlLoop_p->iShouldTerminate)
	{
		ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::start resuming the thread"));
		m_controlLoop_p->iShouldTerminate = false;
	}
	else
	{
		ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::start was already called"));
                SYSTEMErr::SchedulerAlreadyRunningExImpl ex(__FILE__, __LINE__, "called twice");
                throw ex.getSchedulerAlreadyRunningEx();
	}
}

void SchedulerImpl::stop() throw (SYSTEMErr::SchedulerAlreadyStoppedEx)
{
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::stop"));

	// stop the thread
	if(m_controlLoop_p == 0)
	{
		ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::stop thread not created"));
		// TODO: raise an exception
	}
	else if(m_controlLoop_p->iShouldTerminate)
	{
		ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::stop already asked to stop"));
		SYSTEMErr::SchedulerAlreadyStoppedExImpl ex(__FILE__, __LINE__, "called twice");
		throw ex.getSchedulerAlreadyStoppedEx();
	}
	else
	{
		//m_controlLoop_p->suspend();
		m_controlLoop_p->iShouldTerminate = true;
	}
}

CORBA::Long SchedulerImpl::proposalUnderExecution() throw (SYSTEMErr::NoProposalExecutingEx)
{
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::proposalUnderExecution"));
	if(pid_m == -1)
	{
		ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::proposalUnderExecution no proposal under execution"));
		SYSTEMErr::NoProposalExecutingExImpl ex(__FILE__, __LINE__, "no proposal is executing");
		throw ex.getNoProposalExecutingEx();
	}
	else
	{
		return pid_m;
	}
}

void SchedulerImpl::processProposals()
{
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::processProposals"));

	if(getProposalToBeObserved())
	{
		observeProposal();
		pid_m = -1; // proposal has been observed
	}
}

bool SchedulerImpl::getProposalToBeObserved()
{
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::getProposalToBeObserved"));

        database_p = getContainerServices()->getComponent<DATABASE_MODULE::DataBase>("DATABASE");
        ACS_SHORT_LOG((LM_INFO, "SchedulerImpl::getProposalToBeObserved get list of proposals"));
        TYPES::ProposalList_var proposals = database_p->getProposals();
        unsigned int i = 0;
        bool found = false;
        while(i < proposals->length() && !found) {
                ACS_SHORT_LOG((LM_INFO, "SchedulerImpl::processProposals proposal %d", i));
                if(proposals[i].status == QUEUED) {
                        found = true;
                        pid_m = proposals[i].pid;
                        targets_m = proposals[i].targets;
                }
                ++i;
        }
        if(!found)
        {
                ACS_SHORT_LOG((LM_INFO, "SchedulerImpl::getProposalToBeObserved no proposal available to observe"));
                return false;
        }
        ACS_SHORT_LOG((LM_INFO, "SchedulerImpl:getProposalToBeObserved proposal %d available to observe", pid_m));
        try {
                database_p->setProposalStatus(pid_m, RUNNING);
        }
        catch (SYSTEMErr::InvalidProposalStatusTransitionEx &e) {
                ACS_SHORT_LOG((LM_ERROR, "proposal status transition error"));
		pid_m = -1;
                return false;
        }
	return true;
}

void SchedulerImpl::observeProposal()
{
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::observeProposal"));

	instrument_p = getContainerServices()->getComponent<INSTRUMENT_MODULE::Instrument>("INSTRUMENT");
	instrument_p->cameraOn();

	telescope_p = getContainerServices()->getComponent<TELESCOPE_MODULE::Telescope>("TELESCOPE");
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::observeProposal proposal %d: %d targets found", pid_m, targets_m.length()));
	for(unsigned int i = 0; i < targets_m.length(); ++i) {
		int tid = targets_m[i].tid;
		TYPES::ImageType_var image;
		try {
			image = telescope_p->observe(targets_m[i].coordinates, targets_m[i].expTime);
		}
		catch(SYSTEMErr::PositionOutOfLimitsEx &e) {
			ACS_SHORT_LOG((LM_ERROR, "do not get the image for proposal %d target %d", pid_m, tid));
		}
		database_p->storeImage(pid_m, tid, image);
	}

        instrument_p->cameraOff();
        try {
                database_p->setProposalStatus(pid_m, READY);
        }
        catch (SYSTEMErr::InvalidProposalStatusTransitionEx &e) {
                ACS_SHORT_LOG((LM_ERROR, "proposal status transition error"));
                return;
        }
	ACS_SHORT_LOG((LM_INFO,"SchedulerImpl::observeProposal proposal observed"));
}

#include "maciACSComponentDefines.h"
MACI_DLL_SUPPORT_FUNCTIONS(SchedulerImpl)
